compiler/rustc_typeck/src/check/wfcheck.rs

```diff
@@ -863,7 +863,9 @@ fn check_param_wf(tcx: TyCtxt<'_>, param: &hir::GenericParam<'_>) {
                 }
             };
 
-            if traits::search_for_structural_match_violation(param.span, tcx, ty).is_some() {
+            if tcx.features().adt_const_params // structural matching only allowed with adt_const_params feature
```

Comment:
we already have a check for the `adt_const_params` feature a few lines above. I would prefer to only have 1 branch for that. So pretty much

```rust
if tcx.features().adt_const_params {
   // complex check¹
} else {
   // simple check, completely ignoring structural equality
}
```
¹ It might be interesting to check whether function pointers are considered a structural match violation. If they are, you can completely avoid checking the `ty` in that branch.